src/hardware/signal_generator.rs

```diff
@@ -0,0 +1,172 @@
+#[derive(Copy, Clone, Debug)]
+pub enum Signal {
+    Sine,
+    Square,
+    Triangle,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Config {
+    // TODO: Should period be specified in Hz?
+    pub period: u32,
```

Comment:
@jordens In terms of an API over MQTT, would we prefer the user specify the period in Hz, or as a number of samples, thus tied closely to Stabilizer's sampling rate? Currently, it utilizes samples.

Reply:
Internally the frequency should be represented as a "frequency tuning word" i.e. `let ftw = (frequency_hz*((1<<32) as f64/sample_rate_hz)) as i32` and an i32-wrapping accumulator that does `accu += ftw` at each sample (see `Accu`). From that phase you compute the triangle. I suspect it might be best and adequate to restrict the asymmetry to (1-1/(1<<k)):(1/(1<<k)) where k in [1..30]. That should simplify the math.
Note that that frequency and phase (both i32) is then also used in the internal modulation lockin again.
Externally through MQTT it's ok to typically represent as f32 in Hz. But do the conversion in the application, not in the library. Some applications will want to specify them as i32 in machine units (as above).

Reply:
There's an `InternalConf` structure that this config gets converted into that encodes the period as an FTW, so all the math here is integer-oriented with phases represented as i32s. That being said, I didn't limit k to 1..30, but rather all i32s, and then I truncate some LSB data.

Reply:
I know. My point is regarding the way you represent the frequency as a period. That's problematic due to the division inaccuracy. It must be represented as a FTW.

Reply:
I realize now the redundancy in having both a FTW + period register, as if you have a FTW, you can naturally express the period as an overflow of the phase accumulator, so the period register just introduces inaccuracies in phase accumulation. I've refactored this as such, so please let me know if you have further feedback in this regard.